modules/table/src/main/java/org/apache/ignite/internal/table/distributed/disaster/DisasterRecoveryManager.java

```diff
@@ -357,36 +402,94 @@ private static LocalPartitionStateEnum convertState(State nodeState) {
-                return LocalPartitionStateEnum.BROKEN;
+                return BROKEN;
         }
     }
 
     /**
      * Replaces some healthy states with a {@link LocalPartitionStateEnum#CATCHING_UP},it can only be done once the state of all peers is
      * known.
      */
-    private Map<TablePartitionId, Map<String, LocalPartitionState>> normalize(
-            Map<TablePartitionId, Map<String, LocalPartitionState>> result
+    private static Map<TablePartitionId, Map<String, LocalPartitionState>> normalizeLocal(
+            Map<TablePartitionId, Map<String, LocalPartitionStateMessage>> result,
+            Catalog catalog
     ) {
         return result.entrySet().stream().collect(Collectors.toMap(Map.Entry::getKey, entry -> {
-            Map<String, LocalPartitionState> map = entry.getValue();
+            TablePartitionId tablePartitionId = entry.getKey();
+            Map<String, LocalPartitionStateMessage> map = entry.getValue();
 
             // noinspection OptionalGetWithoutIsPresent
-            long maxLogIndex = map.values().stream().mapToLong(LocalPartitionState::logIndex).max().getAsLong();
+            long maxLogIndex = map.values().stream().mapToLong(LocalPartitionStateMessage::logIndex).max().getAsLong();
 
             return map.entrySet().stream().collect(Collectors.toMap(Map.Entry::getKey, entry2 -> {
-                LocalPartitionState state = entry2.getValue();
+                LocalPartitionStateMessage stateMsg = entry2.getValue();
 
-                if (state.state() != LocalPartitionStateEnum.HEALTHY || maxLogIndex - state.logIndex() < CATCH_UP_THRESHOLD) {
-                    return state;
+                LocalPartitionStateEnum stateEnum = stateMsg.state();
+
+                if (stateMsg.state() == HEALTHY && maxLogIndex - stateMsg.logIndex() >= CATCH_UP_THRESHOLD) {
+                    stateEnum = CATCHING_UP;
                 }
 
-                return MSG_FACTORY.localPartitionState()
-                        .state(LocalPartitionStateEnum.CATCHING_UP)
-                        .partitionId(state.partitionId())
-                        .logIndex(state.logIndex())
-                        .build();
+                CatalogTableDescriptor tableDescriptor = catalog.table(tablePartitionId.tableId());
+                return new LocalPartitionState(tableDescriptor.name(), tablePartitionId.partitionId(), stateEnum);
             }));
         }));
     }
+
+    private static Map<TablePartitionId, GlobalPartitionState> normalizeGlobal(
+            Map<TablePartitionId, Map<String, LocalPartitionState>> localResult,
+            Catalog catalog
+    ) {
+        Map<Integer, Integer> tableIdToPartitions = new HashMap<>();
+
+        Map<TablePartitionId, GlobalPartitionState> result = localResult.entrySet().stream()
+                .collect(Collectors.toMap(Map.Entry::getKey, entry -> {
+                    TablePartitionId tablePartitionId = entry.getKey();
```

Comment:
The following block of code looks pretty scary as a lambda. I would suggest to extract it as a method, but filling of `tableIdToPartitions` that is done as a side effect of that lambda makes it more difficult. So, I suggest the following:

1. Do not fill `tableIdToPartitions` in the same block of code; instead, make second iteration over the original map to fill this variable. This aggregation will be extremely rare, so we'll not lose much CPU cycles here.
2. After the lambda does not touch `tableIdToPartitions` anymore, extract it to a method like `assembleGlobalStateFromLocal()`
3. This method and the code for filling `tableIdToPartitions` would reuse the same method to get zone descriptor

It seems that after this refactoring it will become easier to understand this method: it will become much shorter and less coupled.

Reply:
Sure! I don't know if I'll follow your recommendation fully, but I will simplify the code